modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticatorUtils.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.http.netty4;
+
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpRequest;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.http.HttpHeadersValidationException;
+import org.elasticsearch.http.HttpPreRequest;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.getHttpHeadersAsMap;
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.translateRequestMethod;
+
+/**
+ * Provides utilities for hooking into the netty pipeline and authenticate each HTTP request's headers.
+ * See also {@link Netty4HttpHeaderValidator}.
+ */
+public final class HttpHeadersAuthenticatorUtils {
+
+    /**
+     * An async HTTP headers authenticating function that receives as arguments part of the incoming HTTP request
+     * (except the body contents, see {@link HttpPreRequest}), as well as the netty channel that the request is
+     * being received over, and must then call the {@code ActionListener#onResponse} method on the listener parameter
+     * in case the validation is to be considered successful, or otherwise call {@code ActionListener#onFailure}
+     * and pass the failure exception.
+     */
+    @FunctionalInterface
+    public interface Authenticator extends TriConsumer<HttpPreRequest, Channel, ActionListener<Void>> {}
```

Comment:
There are two uses of `TriConsumer` in the code:
1. The parameter to `Netty4HttpHeaderValidator`, which is what we call a "validator". This one takes a netty type as a "request" parameter. Netty types are not exported by the `org.elasticsearch.transport.netty4` module.
2. The parameter to `HttpHeadersAuthenticatorUtils#getValidatorInboundHandler`, which is what we call an "authenticator". This one takes a type from the `org.elasticsearch.server` module as the "request" parameter.

The 2nd point exists mainly because the `org.elasticsearch.security` module needs to ultimately provide an implementation for the "validator", but it cannot access the netty types.

That being said, I have pushed 2 PRs, for the corresponding 2 points above to remove the usage of the `TriConsumer`:
1. [d2b363d](https://github.com/elastic/elasticsearch/pull/95112/commits/d2b363d35f23857e6401ab37f6d005c8aa447530) adds an interface type that is not exportable outside the `org.elasticsearch.transport.netty4` module.
2. [c7eeba1](https://github.com/elastic/elasticsearch/pull/95112/commits/c7eeba118ecc0f6e5e8048985ae78a572a46c4f4) adds another interface type that is only exported to the `org.elasticsearch.security` module.

Reply:
This is looking good, however I don't think we need the `HttpAuthenticator` interface and can use only 1 interface `HttpValidator`. Having the authentication in the name is helpful but having 2 near identical feel abit off and the authentication wording discussed prior is mostly covered by the other aptly named classes and methods. 

To do this we will need to need to add requires to the security module, i.e. in x-pack/plugin/security/src/main/java/module-info.java add the following 
```java
requires io.netty.codec.http;
```
Then we can make `asHttpPreRequest(HttpRequest request)` public and call it from the security plugin and remove references to `HttpAuthenticator` inteface i.e.
```java
.getValidatorInboundHandler((httpRequest, channel, listener) -> {
                    // step 1: Populate the thread context with credentials and any other HTTP request header values (eg run-as) that the
                    // authentication process looks for while doing its duty.
                    HttpPreRequest httpPreRequest = HttpHeadersAuthenticatorUtils.asHttpPreRequest(httpRequest); // <-- convert here
                    perRequestThreadContext.accept(httpPreRequest, threadContext);
```
Finally we can move `HttpHeadersWithAuthenticationContext` and `HttpHeadersAuthenticatorUtils` to the internal package (not technically needed but helps to keep all the internal pieces together and private to the security plugin) and update the exports to look like:
```java
   exports org.elasticsearch.http.netty4.internal to org.elasticsearch.security;
```

Generally we don't want to export something we can just omit it instead of added a export to a non-exiting module.

Reply:
Adoption of the above ^^ could also slightly simplify the contract to `Netty4HttpServerTransport` such that with little effort we don't supply the entire netty header validator: ` @Nullable Supplier<Netty4HttpHeaderValidator> headersValidatorSupplier` but instead supply the sole `HttpValidator` and have a similar null check to do validation or not, but just null check for the `HttpValidator` supplier instead of Netty4HttpHeaderValidator.  It feels like the Netty4HttpHeaderValidator is closer to framework/infrastructure code and `HttpValidator` is more business logic code and as such IMO it makes a bit more sense for the `HttpValidator` to be the contract between these modules.

Reply:
Pushed [b524b7c](https://github.com/elastic/elasticsearch/pull/95112/commits/b524b7c38de820600374c623b6ef472234d0fdf1) and [2b73b71](https://github.com/elastic/elasticsearch/pull/95112/commits/2b73b71a0509c9c06c1cb38f17ecc6e78a27e78b) .